lib/Audio/SunVox/FFI/Module.pm

```diff
@@ -0,0 +1,339 @@
+    } );
+}
+
+sub _ctl {
+    my ( $ctl, $scale ) = @_;
+    my ( $min, $max, $method_name ) = @{ $ctl }{ "min_$scale", "max_$scale", 'method_name' };
+    sub {
+        my ( $self, $value ) = @_;
+
+        return sv_get_module_ctl_value( $self->slot->num, $self->num, $ctl->{ ctl_num }, $self->scale( $scale ) )
+            unless defined $value;
+
+        goto nobounds if $self->skip_bounds_checking;
+        if ( $value < $min ) {
+            carp "Value $value below minimum of $min for $ctl->{ method_name } - setting to $min";
+            $value = $min;
+        }
+        elsif ( $value > $max ) {
+            carp "Value $value below maximum of $max for $ctl->{ method_name } - setting to $max";
+            $value = $max;
+        }
+nobounds:
+        sv_set_module_ctl_value(
+            $self->slot->num,
+            $self->num,
+            $ctl->{ ctl_num },
+            $value,
+            $self->scale( $scale )
+        );
+    }
+}
+
+sub import {
+    my ( $pkg, %cfg ) = @_;
+    $default_scale = $cfg{ default_scale } if $cfg{ default_scale };
+    __PACKAGE__->export_to_level( 1, @_ );
+}
+
+sub module {
+    my ( $slot, $type, $name ) = @_;
+    my $class = $module_data->{ $type }->{ class_name };
+    croak "Unknown module type : $type" unless $class;
+    $slot //= Audio::SunVox::FFI::Slot->get_last;
+    $class->new(
+        slot => $slot,
+        name => $name,
+        type => $type,
+    );
+}
+
+sub new {
+    my ( $class, %params ) = @_;
+    if ( $class eq 'Audio::SunVox::FFI::Module' ) {
+        carp "Audio::SunVox::FFI::Module should not be instantiated directly";
+        return -1;
+    }
+    $params{ slot } //= Audio::SunVox::FFI::Slot->get_last;
+    my $self = bless \%params, $class;
+    $self->add_to_slot( $params{ name } ) unless $self->{ in_slot };
+    $self;
+}
+
+sub skip_bounds_checking {
+    my ( $self, $val ) = @_;
+    $self->{ skip_bounds_checking } = $val if defined $val;
+    $self->{ skip_bounds_checking };
+}
+
+sub scale { $scales->{ $_[1] } }
+
+sub num { shift->{ num } }
+
+sub slot { shift->{ slot } }
+
+sub default_scale { shift->scale( $default_scale ) }
+
+sub connect_to {
+    my ( $self, @modules ) = @_;
+    $self->slot->lock;
+    sv_connect_module( $self->slot->num, $self->num, $modules[0]->num );
+    $self->slot->unlock;
+    return ( $self, @modules );
+}
+*connect = \&connect_to;
+
+sub disconnect {
+    my ( $self, $module ) = @_;
+    sv_disconnect_module( $self->num, $module->num );
+}
+
+sub send_event {
+    my ( $self, $track, $note, $vel, $ctl, $val ) = @_;
+    $ctl //= 0;
+    sv_send_event( $self->slot->num, $track, $note, $vel, $self->num + 1, $ctl << 8, $val );
+}
+
+sub note_on {
+    my ( $self, $track, $note, $vel ) = @_;
+    sv_send_event( $self->slot->num, $track, $note, $vel, $self->num + 1 );
+}
+
+sub note_off {
+    my ( $self, $track, $note ) = @_;
+    sv_send_event( $self->slot->num, $track, NOTECMD_NOTE_OFF, $note, $self->num + 1 );
+}
```

Comment:
This interface is lacking - it's treating events like MIDI. The note off event doesn't apply to a module or any playing note, it applies to the *track*. That is, any module can tell any track to stop playing its current note.

Consider assigning a realtime track number to a module on construction, and limiting these convenience methods to that track. Also think about polyphony.

The `note_on` method should allow for other columns - FX, parameters and so on. (Maybe `note_off` should too?)

Further, think about where `send_event` belongs... Maybe multiple context-sensitive versions of this are required.

Reply:
Consider also : mobility. Cloning a module between slots? Should be possible to copy attributes and curves, at least ... connections directly to Output should be retained, otherwise all bets are off.